File: benchmark/interactive_function/prisoners/prisoners.java
import java.util.concurrent.ThreadLocalRandom;

public class prisoners{
    public static double playOptimal(int n) {
        int pardoned = 0;
        int[] inDrawer = new int[100];
        for (int i = 0; i < 100; i++) inDrawer[i] = i;

        for (int r = 0; r < n; r++) {
            // Shuffle
            for (int i = 99; i > 0; i--) {
                int index = ThreadLocalRandom.current().nextInt(i + 1);
                int a = inDrawer[index];
                inDrawer[index] = inDrawer[i];
                inDrawer[i] = a;
            }

            boolean allFound = true;
            for (int prisoner = 0; prisoner < 100; prisoner++) {
                boolean found = false;
                int reveal = prisoner;
                for (int go = 0; go < 50; go++) {
                    if (inDrawer[reveal] == prisoner) {
                        found = true;
                        break;
                    }
                    reveal = inDrawer[reveal];
                }
                if (!found) {
                    allFound = false;
                    break;
                }
            }
            if (allFound) pardoned++;
        }
        return (double) pardoned / n * 100;
    }

    public static void main(String[] args) {
        int n = 1000000;
      System.out.printf("Optimal play wins (Java): %.1f%%\n", playOptimal(n));
    }
}

File: benchmark/interactive_function/prisoners/prisoners.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

func playOptimal(n int) float64 {
	pardoned := 0
	inDrawer := make([]int, 100)
	for i := 0; i < 100; i++ {
		inDrawer[i] = i
	}

	for r := 0; r < n; r++ {
		rand.Shuffle(100, func(i, j int) { inDrawer[i], inDrawer[j] = inDrawer[j], inDrawer[i] })

		allFound := true
		for prisoner := 0; prisoner < 100; prisoner++ {
			found := false
			reveal := prisoner
			for go_round := 0; go_round < 50; go_round++ {
				card := inDrawer[reveal]
				if card == prisoner {
					found = true
					break
				}
				reveal = card
			}
			if !found {
				allFound = false
				break
			}
		}
		if allFound {
			pardoned++
		}
	}
	return float64(pardoned) / float64(n) * 100
}

func main() {
	rand.Seed(time.Now().UnixNano())
	n := 1000000
	fmt.Printf("Optimal play wins (Go): %.1f%%\n", playOptimal(n))
}
